core/src/main/java/org/elasticsearch/search/fetch/subphase/DocValueFieldsContext.java

```diff
@@ -18,23 +18,137 @@
+                return new Field(parser.text(), null);
+            } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                String name = null;
+                String format = null;
+                String currentFieldName = null;
+                XContentParser.Token token;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        currentFieldName = parser.currentName();
+                    } else if (context.getParseFieldMatcher().match(currentFieldName, DOCVALUE_FIELD_NAME)) {
+                        name = parser.text();
+                    } else if (context.getParseFieldMatcher().match(currentFieldName, DOCVALUE_FIELD_FORMAT)) {
+                        format = parser.textOrNull();
+                    } else {
+                        throw new ParsingException(parser.getTokenLocation(), "Unknown property under ["
+                                + DOCVALUE_FIELDS_FIELD.getPreferredName() + "] : [" + currentFieldName
+                                + "]", parser.getTokenLocation());
+                    }
+                }
+                if (name == null) {
+                    throw new ParsingException(parser.getTokenLocation(), "Missing name for ["
+                            + DOCVALUE_FIELDS_FIELD.getPreferredName() + "]", parser.getTokenLocation());
+                }
+                return new Field(name, format);
+            } else {
+                throw new ParsingException(parser.getTokenLocation(), "Expected [" + XContentParser.Token.VALUE_STRING +
+                        "] or [" + XContentParser.Token.START_OBJECT + "] in [" + DOCVALUE_FIELDS_FIELD.getPreferredName() +
+                        "] but found [" + parser.currentToken() + "]", parser.getTokenLocation());
+            }
+        }
+
+        private final String name;
+        private final String format;
+
+        public Field(String name, String format) {
+            this.name = Objects.requireNonNull(name);
+            this.format = format;
+        }
+
+        public Field(StreamInput in) throws IOException {
+            this(in.readString(), in.readOptionalString());
+        }
+
+        @Override
+        public String toString() {
+            return "Field(name=" + name + ", format=" + format + ")";
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null || getClass() != obj.getClass()) {
+                return false;
+            }
+            Field other = (Field) obj;
+            return Objects.equals(name, other.name) && Objects.equals(format, other.format);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(name, format);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(name);
```

Comment:
Can you move this up under the reading constructor? It helps to have them on the same screen.